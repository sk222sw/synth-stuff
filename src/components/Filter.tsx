import * as React from "react";
import Knob from "./Knob";

interface Props {
  type: string;
  frequency: number;
  onFrequencyChange: (value: number) => void;
}

const Filter = ({ type, frequency, onFrequencyChange }: Props) => (
  <div>
    <div>{type}</div>
    <Knob max={20000} onChange={onFrequencyChange}></Knob>
  </div>
);

export default Filter;
