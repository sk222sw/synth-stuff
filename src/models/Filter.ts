export interface IFilterConfig {
  frequency: number
  type: BiquadFilterType
}

export interface IFilter extends IFilterConfig {
  filterNode?: BiquadFilterNode,
}

const createFilter = (audioContext: AudioContext, { frequency = 7000, type = 'lowpass' } = {}): IFilter => {
  const filterNode = audioContext.createBiquadFilter()
  const _type = type as BiquadFilterType;
  filterNode.type = _type 
  ;filterNode.frequency.value = frequency

  return {
    frequency,
    type: _type,
    filterNode,
  }
}

const setFilterType = (filter: IFilter, type: string) => {
  if (!filter.filterNode) throw new Error('No filter node')
  const _type = type as BiquadFilterType;
  filter.filterNode.type = _type 
  filter.type = _type

  return filter
}

const setFrequency = (filter: IFilter, frequency: number) => {
  if (!filter.filterNode) throw new Error('No filter node')
  filter.filterNode.frequency.value = frequency
  filter.frequency = frequency

  return filter
}

const Filter = {
  createFilter,
  setFilterType,
  setFrequency,
}

export default Filter