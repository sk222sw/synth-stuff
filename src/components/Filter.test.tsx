import { shallow } from 'enzyme'
import * as React from 'react'
import Filter from './Filter'

test('Filter', () => {
  const spy = jest.fn()

  const wrapper = shallow(
    <Filter
      frequency={200}
      onFrequencyChange={spy}
      type={'lowpass'}
    />,
  )

  expect(wrapper).toMatchSnapshot()
})
