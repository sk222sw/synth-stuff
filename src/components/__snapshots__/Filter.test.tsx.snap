// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <div>
        lowpass
    </div>
    <Knob
        angleArc={250}
        angleOffset={-125}
        bgColor="#EEE"
        clockwise={true}
        cursor={false}
        disableMouseWheel={false}
        disableTextInput={false}
        displayInput={true}
        fgColor="#EA2"
        font="Arial"
        fontWeight="bold"
        height={50}
        inputColor=""
        lineCap="butt"
        log={false}
        max={20000}
        min={0}
        onChange={[Function]}
        onChangeEnd={[Function]}
        readOnly={false}
        step={1}
        stopper={true}
        thickness={0.3}
        value={200}
        width={50}
    />
</div>,
  "nodes": Array [
    <div>
      <div>
            lowpass
      </div>
      <Knob
            angleArc={250}
            angleOffset={-125}
            bgColor="#EEE"
            clockwise={true}
            cursor={false}
            disableMouseWheel={false}
            disableTextInput={false}
            displayInput={true}
            fgColor="#EA2"
            font="Arial"
            fontWeight="bold"
            height={50}
            inputColor=""
            lineCap="butt"
            log={false}
            max={20000}
            min={0}
            onChange={[Function]}
            onChangeEnd={[Function]}
            readOnly={false}
            step={1}
            stopper={true}
            thickness={0.3}
            value={200}
            width={50}
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Filter
        frequency={200}
        onFrequencyChange={[Function]}
        type="lowpass"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Filter {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "frequency": 200,
          "onFrequencyChange": [Function],
          "type": "lowpass",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <div>
                    lowpass
          </div>
          <Knob
                    angleArc={250}
                    angleOffset={-125}
                    bgColor="#EEE"
                    clockwise={true}
                    cursor={false}
                    disableMouseWheel={false}
                    disableTextInput={false}
                    displayInput={true}
                    fgColor="#EA2"
                    font="Arial"
                    fontWeight="bold"
                    height={50}
                    inputColor=""
                    lineCap="butt"
                    log={false}
                    max={20000}
                    min={0}
                    onChange={[Function]}
                    onChangeEnd={[Function]}
                    readOnly={false}
                    step={1}
                    stopper={true}
                    thickness={0.3}
                    value={200}
                    width={50}
          />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <div>
                    lowpass
          </div>
          <Knob
                    angleArc={250}
                    angleOffset={-125}
                    bgColor="#EEE"
                    clockwise={true}
                    cursor={false}
                    disableMouseWheel={false}
                    disableTextInput={false}
                    displayInput={true}
                    fgColor="#EA2"
                    font="Arial"
                    fontWeight="bold"
                    height={50}
                    inputColor=""
                    lineCap="butt"
                    log={false}
                    max={20000}
                    min={0}
                    onChange={[Function]}
                    onChangeEnd={[Function]}
                    readOnly={false}
                    step={1}
                    stopper={true}
                    thickness={0.3}
                    value={200}
                    width={50}
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Filter
    frequency={200}
    onFrequencyChange={[Function]}
    type="lowpass"
/>,
}
`;
